synapse/lib/blobfile.py

```diff
@@ -0,0 +1,310 @@
+
+    def syncs(self, msgs):
+        '''
+        Consume a list of blob:sync events.
+
+        Args:
+            msgs (list): A list of blob:sync messages.
+
+        Returns:
+            None
+        '''
+        [self.syncact.react(mesg[1].get('mesg')) for mesg in msgs]
+
+    def _fireBlobSync(self, mesg):
+        self.fire('blob:sync', mesg=mesg)
+
+    def _actSyncBlobWrite(self, mesg):
+        # event is triggered *with* fdlock
+        off = mesg[1].get('off')
+        byts = mesg[1].get('byts')
+        self._writeoff(off, byts)
+
+    def _actSyncBlobAlloc(self, mesg):
+        size = mesg[1].get('size')
+        self.alloc(size)
+
+    def _writeoff(self, off, byts):
+        self.atom.writeoff(off, byts)
+        self.fire('blob:write', off=off, byts=byts)
+
+    def readoff(self, off, size):
+        '''
+        Read and return bytes from the blob at an offset.
+
+        Args:
+            off (int): Offset to read from.
+            size (int): Number of bytes to read.
+
+        Examples:
+            Read the blobsize of the first blob stored at in the blobfile:
+
+                head = blob.readoff(0,s_blobfile.headsize)
+
+        Returns:
+            bytes: The bytes from a given offset.
+
+        Raises:
+            s_common.BadBlobFile: If not enough bytes were read to fullfill the request.
+        '''
+        byts = self.atom.readoff(off, size)
+
+        if len(byts) != size:
+            raise s_common.BadBlobFile(mesg='readoff was short',
+                                       expected_size=size,
+                                       byts_len=len(byts))
+
+        return byts
+
+    def readiter(self, off, size, itersize=10000000):
+        '''
+        Yield back bytes chunks for the given offset and size.
+
+        Args:
+            off (int): Offset to read from.
+            size (int): Number of bytes to read.
+            itersize (int): Maximum number of bytes to yield in a chunk.
+
+
+        Examples:
+            Call dostuff() on byts as they are read from the blob:
+
+                for byts in blob.readiter(off,size):
+                    dostuff(byts)
+
+        Yields:
+            bytes: Chunks of bytes.
+        '''
+        offmax = off + size
+
+        while off < offmax:
+
+            offend = min(off + itersize, offmax)
+
+            yield self.atom.readoff(off, offend - off)
+
+            off = offend
+
+    def writeoff(self, off, byts):
+        '''
+        Write bytes at an offset within the blob.
+
+        Args:
+            off (int): Offset to write bytes too.
+            byts (bytes): Bytes to write at the offset.
+
+        Examples:
+            Allocate space for bytes and write it to the blob:
+
+                byts = 'hehe'.encode()
+                sz = len(byts)
+                offset = blob.alloc(sz)
+                blob.writeoff(offset, byts)
+
+        Returns:
+            None
+        '''
+        self._writeoff(off, byts)
```

Comment:
The alloc/writeoff APIs should *not* be callable if we are a blob sync event consumer since the write events are assuming that all alloc's got the same offsets in the consumer.  I'd recommend a local  self.isclone that gets set on startup and checked for any of the write APIs :)

Reply:
we can plumb that in super easily

Reply:
plumb